### DISCUSSION TITLE
What is the diffrence between `noblacklist` and `whitelist`?

### DISCUSSION THREAD

**Discussion by rusty-snake** [netblue30/firejail#4247] [Q&A] [answered]
> Also one more query i had was what is diffrence between `noblacklist` and `whitelist`.
> if i use `noblacklist ${HOME}/.vim` followed by `whitelist ${HOME}/.config/joplin` then `~/.vim` doesn't show up in the sandbox.
> if i use `noblacklist` both places then all elements in home directory are visible for the app (including `~/my_personal_folder`).
> if i use `whitelist` both place then only .vim and required config shows up.

https://github.com/netblue30/firejail/issues/1139#issuecomment-832688052 by @sak96

**Comment by rusty-snake** [comment, marked as answer]
### `noblacklist`

`noblacklist ${HOME}/foo/bar` makes that all *following* `blacklist ${HOME}/foo/bar` commands are ignored.
It is, if you will, only syntactic-sugar for `ignore blacklist ${HOME}/foo/bar`.

 - Having a `noblacklist ${HOME}/foo/bar` without `blacklist ${HOME}/foo/bar` in a `include disable-*.inc` is just useless.
 - `noblacklist ${HOME}/foo/bar` and `blacklist ${HOME}/foo` does *not* work. You can only `noblacklist` a path that is `blacklist`e
d.

### `whitelist`

`whitelist ${HOME}/foo/bar` enables `whitelist`ing in `${HOME}` (i.e. everything without a explicit whitelist is hidden and discard on close) and makes `${HOME}/foo/bar` and persistent.

 - `blacklist ${HOME}/foo/bar` is still applied, even with `whitelist ${HOME}/foo/bar`.

**Comment by rusty-snake** [comment]
> if i use noblacklist ${HOME}/.vim followed by whitelist ${HOME}/.config/joplin then ~/.vim doesn't show up in the sandbox.

because of missing `whitelist` in a whitelisting situation

> if i use noblacklist both places then all elements in home directory are visible for the app (including ~/my_personal_folder).

no `whitelist`, no whitelisting

> if i use whitelist both place then only .vim and required config shows up.

whitelisting enabled, everything else hidden

**Comment by sak96** [comment]
how is the depth determined in whitelist 
example /usr/share/z/y will hide all thing in /usr ?
can it be made to only hide /usr/share contents ?

**Reply by rusty-snake** [reply]
> **--whitelist=dirname_or_filename**
> Whitelist directory or file. A temporary file system is mounted on the top directory, and the
> whitelisted files are mount-binded inside. Modifications to whitelisted files are persistent,
> everything else is discarded when the sandbox is closed. The top directory could be
> user home, /dev, /etc, /media, /mnt, /opt, /run/user/$UID, /srv, /sys/module, /tmp, /usr/share and /var.
